### ISSUE TITLE
Modify static transpilation of for loops into `__createKernelSource` calls

### ISSUE THREAD

**Issue by limhawjia** [limhawjia/js-slang#5]
Now that the GPU library primitives are successfully modified, it is time to modify the static transpilation of for-loops.

The existing implementation performs the following steps:

* Verify the for statements
* Verify the for loop body
* Transform for loops into a function call

Our new transpilation should extract the extra information required, such as the counter names and indices used in array assignment, and pass them to the new primitive functions.

The current implementation also has a bug that occurs if we have more than 3 counters being used as indices. For example, the following program:

```javascript
for (let i = 0; i < 5; i = i + 1) {
    for (let j = 0; j < 2; j = j + 1) {
        for (let k = 0; k < 2; k = k + 1) {
            for (let l = 0; l < 3; l = l + 1) {
                arr[i][j][k][l] =  2;
            }
        }
    }
}
```

Results in the following incorrect results:

```javascript
"Attempting to optimize 3 levels of nested loops starting on line 61"

[ [[2, 2], [2, 2]],
  [[2, 2], [2, 2]],
  [[2, 2], [2, 2]],
  [[2, 2], [2, 2]],
  [[2, 2], [2, 2]]]
```

Correct answer should be:

```javascript
[ [[[2, 2, 2], [2, 2, 2]], [[2, 2, 2], [2, 2, 2]]],
  [[[2, 2, 2], [2, 2, 2]], [[2, 2, 2], [2, 2, 2]]],
  [[[2, 2, 2], [2, 2, 2]], [[2, 2, 2], [2, 2, 2]]],
  [[[2, 2, 2], [2, 2, 2]], [[2, 2, 2], [2, 2, 2]]],
  [[[2, 2, 2], [2, 2, 2]], [[2, 2, 2], [2, 2, 2]]]]
```

The error is caused because only the 3 outer-most for loops are parallelized into threads. Any inner loops are taken as part of the function body to be passed to GPU.js.

In the function body, the array assignment is stripped away, the RHS is converted to a return value. Whatever is returned by GPU.js is copied into the target array.

We can still do parallellization of 3 loops at a time, but we must do it for the 3 inner most loops. We should still iterate through the values of the outer loops sequentially and copy the results of the GPU.js function to the corresponding coordinates.

The new implementation should account for this.

**Comment by limhawjia** [comment]
Step 1: modify checks to accommodate our new specifications

* Loop checks can be left untouched, our new specifications only loosen the restrictions on the array indices being used.
* Previously, in `GPUBodyVerifier`, the indices are checked to see if they are a prefix of the loop counters. Now, we allow any combination of indices, as long as they are either a counter, an external variable, or a number.

Previously, information about the number of indices, local variables and the output array is retrieved in the process of checking the loop body. Now, we need to extract more information, namely the indices used in the array assignment.

**Comment by limhawjia** [comment]
Step 2: modify transpilation of loop into function call

The previous implementation had some bugs because this was done wrongly. In the event that there were more than 3 indices, the extra inner for statements were treated as part of the kernel function body to be passed to GPU.js. This resulted in incorrect behavior.

Rather, what should be done is that we only use the inner indices for parallelization, and for any further indices that counters, we simply keep the corresponding outer loop, and we assign to the corresponding subarray.

For example in the following program:

```javascript
for (let i = 0; i < 2; i = i + 1) {
    for (let j = 0; j < 3; j = j + 1) {
        for (let k = 0; k < 4; k = k + 1) {
            for (let l = 0; l < 4; l = l + 1) {
                arr[i][j][k][l] = i + j + k + l;
            }
        }
    }
}
```

It should be transpiled to:

```javascript
for (let i = 0; i < 2; i = i + 1) {
    const __arr = arr[i];
    __createKernelSource(...., _arr, ...);
}
```

Looks like we have to reserve another keyword `__arr` in addition to the GPU primitive functions.

**Comment by limhawjia** [comment]
Completed, left with unit tests for `gpuTranspile`.

**Comment by limhawjia** [comment]
Unit tests completed, resolved with pr #6
